### PR TITLE
feat(flat-table): Migrate dls table styles (FE-2690)

### PR DESCRIPTION
**Pull request by grabkowski**
### Proposed behaviour
Implement these `Table` component functionalities in the `FlatTable`:
- zebra stripes,
- table caption,
- passing `aria-describedby` to the `table` element,
- size prop to control various predefined cell sizes.

Move color themes mdx to a separate storybook folder.

Fix an issue with `FlatTableHeaderRow` not being highlighted on click.

### Checklist

- [x] Commits follow our style guide
~~- [ ] Screenshots are included in the PR if useful~~
~~- [ ] All themes are supported if required~~
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Carbon implementation and Design System documentation are congruent

### Testing instructions

1. npm start.
2. sizes: http://localhost:9001/?path=/docs/design-system-flat-table-sizes--compact
3. color themes: http://localhost:9001/?path=/docs/design-system-flat-table-color-themes--transparent-white-theme
4. caption and aria-describedby: http://localhost:9001/?path=/story/design-system-flat-table-test--basic
5. zebra: http://localhost:9001/?path=/docs/design-system-flat-table--zebra

### CODE REVIEW

File: src/components/flat-table/flat-table.component.js [hunk 1]
```diff
@@ -9,15 +9,28 @@ import { SidebarContext } from "../drawer";
...
   colorTheme,
   footer,
   hasStickyFooter = false,
   height,
+  isZebra,
+  size,
   ...props
 }) => {
   const addDefaultHeight = !height && (hasStickyHead || hasStickyFooter);
+  const tableStylingProps = {
```
[the remark this answers is not in the document]
Comment (grabkowski) [reply src/components/flat-table/flat-table.component.js:24]: you need to conditionally add `aria-describedby` to that object, please check line 30.

File: src/components/flat-table/flat-table.style.js [hunk 2]
```diff
@@ -1,18 +1,74 @@
...
+        ${StyledFlatTableRowHeader},
+        ${StyledFlatTableCell},
+      ${StyledFlatTableCheckbox} {
+          background-color: ${theme.table.zebra};
+        }
+      }
+      ${StyledFlatTableRow}:hover {
+        ${StyledFlatTableCell},
+        ${StyledFlatTableRowHeader},
+      ${StyledFlatTableCheckbox} {
```
[the remark this answers is not in the document]
Comment (grabkowski) [reply src/components/flat-table/flat-table.style.js]: fixed

File: src/components/flat-table/flat-table.style.js [hunk 3]
```diff
@@ -1,18 +1,74 @@
...
+    `;
+  }}
+
+  ${({ isZebra, theme }) =>
+    isZebra &&
+    css`
+      ${StyledFlatTableRow}:nth-child(2n) {
+        ${StyledFlatTableRowHeader},
+        ${StyledFlatTableCell},
+      ${StyledFlatTableCheckbox} {
```
[the remark this answers is not in the document]
Comment (grabkowski) [reply src/components/flat-table/flat-table.style.js]: fixed

File: src/components/flat-table/flat-table.style.js [hunk 4]
```diff
@@ -1,18 +1,74 @@
...
+
+  ${({ caption }) =>
+    caption &&
+    css`
+      caption {
+        clip: rect(1px, 1px, 1px, 1px);
+        height: 1px;
+        overflow: hidden;
+        width: 1px;
+        position: absolute;
```
[the remark this answers is not in the document]
Comment (grabkowski) [reply src/components/flat-table/flat-table.style.js:26]: I think that we'll end up with a 1px white line in the table

File: src/components/flat-table/flat-table-sizes.stories.mdx [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+  FlatTableCell,
+} from '.';
+
+<Meta title="Design System/Flat Table/Sizes" parameters={{ info: { disable: true }}} />
+
+## Table Cell Sizes
+
+### compact
+
+<Preview>
```
[the remark this answers is not in the document]
Comment (grabkowski) [reply src/components/flat-table/flat-table-sizes.stories.mdx:16]: good point

File: src/components/flat-table/flat-table-themes.stories.mdx [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+  FlatTableCell,
+} from '.';
+
+<Meta title="Design System/Flat Table/Color Themes" parameters={{ info: { disable: true }}} />
+
+## Table Header Color Themes
+
+### With transparent-white theme
+To create a transparent effect for the header, the background-color is set to match a white background
+
```
[the remark this answers is not in the document]
Comment (grabkowski) [reply src/components/flat-table/flat-table-themes.stories.mdx:17]: we have descriptions on these stories and we'll lose them when we iterate that

**Review by mkrds — APPROVED** [review]

File: src/components/flat-table/flat-table-themes.stories.mdx [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+          <FlatTableCell>Newcastle</FlatTableCell>
+          <FlatTableCell>Married</FlatTableCell>
+          <FlatTableCell>5</FlatTableCell>
+        </FlatTableRow>
+      </FlatTableBody>
+    </FlatTable>
+  </Story>
+</Preview>
+
+### With light theme
```
Comment (DipperTheDan) [inline src/components/flat-table/flat-table-themes.stories.mdx:106]: The other two examples have descriptions but this one doesn't. Do you think this example would benefit from having a description too?

Comment (grabkowski) [reply src/components/flat-table/flat-table-themes.stories.mdx:106]: honestly I didn't thought about that as I just moved these stories so they are being kept together

**Review by DipperTheDan — DISMISSED** [review]

File: src/components/flat-table/flat-table-sizes.stories.mdx [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  <Story name="compact">
+    <div>
+      {["compact", "small", "medium", "large"].map(size => <Box mb="3">
+        <FlatTable size={size}>
+          <FlatTableHead>
+            <FlatTableRow>
+              <FlatTableHeader>Name</FlatTableHeader>
+              <FlatTableHeader>Location</FlatTableHeader>
+              <FlatTableHeader>Relationship Status</FlatTableHeader>
+              <FlatTableHeader>Dependents</FlatTableHeader>
```
Comment (Alec-Walker-Sage) [inline src/components/flat-table/flat-table-sizes.stories.mdx:26]: Can we add or change a row to show an input box. We can then show that the input box changes size to correspond to the DLS

File: src/components/flat-table/flat-table.component.js [hunk 9]
```diff
@@ -21,6 +21,16 @@ const FlatTable = ({
...
+    caption,
+    isZebra,
+    size,
+  };
+
+  if (props["aria-describedby"]) {
```
Comment (Alec-Walker-Sage) [inline src/components/flat-table/flat-table.component.js:30]: described by is also added to a div that contains the table 
image is this intentional?

File: src/components/flat-table/flat-table-sizes.stories.mdx [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+  FlatTableHeader,
+  FlatTableCell,
+} from '.';
+
+<Meta title="Design System/Flat Table/Sizes" parameters={{ info: { disable: true }}} />
+
+## Table Cell Sizes
+
+<Preview>
+  <Story name="compact">
```
Comment (Alec-Walker-Sage) [inline src/components/flat-table/flat-table-sizes.stories.mdx:17]: Story name should not be compact. Should be something like "sizes" as it shows all the sizes. Or might be worth removing the sizes folder and placing this story alongside the others in the Design System/Flat Table folder

**Review by Alec-Walker-Sage — CHANGES_REQUESTED** [review]

File: src/components/flat-table/flat-table-sizes.stories.mdx (hunk 8 above)
Comment (grabkowski) [reply src/components/flat-table/flat-table-sizes.stories.mdx]: input boxes should be sized separately as they have their own `size` prop

File: src/components/flat-table/flat-table-sizes.stories.mdx (hunk 10 above)
Comment (grabkowski) [reply src/components/flat-table/flat-table-sizes.stories.mdx]: changed to `sizes` and moved to common folder

File: src/components/flat-table/flat-table.component.js (hunk 9 above)
Comment (grabkowski) [reply src/components/flat-table/flat-table.component.js]: good catch, this prop was just passed with all other props. Removed.

File: cypress/features/regression/designSystem/flatTable.feature [hunk 11]
```diff
@@ -84,4 +84,19 @@ Feature: Design Systems FlatTable component
     Examples:
       | headerName | position |
       | Client     | first    |
-      | total      | second   |
\ No newline at end of file
+      | total      | second   |
+    
+  @posative
```
Comment (DipperTheDan) [inline cypress/features/regression/designSystem/flatTable.feature:89]: Spelling mistake here. Should be `@Positive`.

File: src/components/flat-table/flat-table.stories.mdx [hunk 12]
```diff
@@ -1142,6 +1050,53 @@ of the button control.
   </Story>
 </Preview>
 
+## Table Cell Sizes
+
+<Preview>
+  <Story name="sizes">
+    <div>
+      {["compact", "small", "medium", "large"].map(size => <Box mb="3">
```
Comment (Alec-Walker-Sage) [inline src/components/flat-table/flat-table.stories.mdx:1058]: There is no margin, I tried editing the mb prop but couldn't figure it out 
image

File: cypress/features/regression/designSystem/flatTable.feature (hunk 11 above)
Comment (Alec-Walker-Sage) [reply cypress/features/regression/designSystem/flatTable.feature]: My speling has been bad lattely

**Review by Alec-Walker-Sage — DISMISSED** [review]

File: src/components/flat-table/flat-table.stories.mdx (hunk 12 above)
Comment (grabkowski) [reply src/components/flat-table/flat-table.stories.mdx]: looks like Box was not imported here, it should work now

File: cypress/locators/flat-table/index.js [hunk 13]
```diff
@@ -11,7 +11,10 @@ export const flatTable = () => cy.get(FLAT_TABLE_COMPONENT);
 export const flatTableHeader = () => flatTable().find('thead tr');
 export const flatTableHeaderCells = () => flatTableHeader().find('th');
 export const flatTableBodyRowByPosition = index => flatTable().find('tbody tr').eq(index);
+export const flatTableBodyRows = () => flatTable().find('tbody tr');
 
 export const flatTableClickableRow = index => cy.get(FLAT_TABLE_COMPONENT).find('tbody tr').eq(index);
 export const flatTableSortable = () => cy.get(FLAT_TABLE_COMPONENT).find('thead tr th div [type=button]');
 export const flatTableCell = index => cy.get(FLAT_TABLE_CELL).eq(index);
+
+export const flatTableCaption = () => cy.get(FLAT_TABLE_COMPONENT).find('caption');
```
Comment (KatarzynaQA) [inline cypress/locators/flat-table/index.js:20]:
```suggestion
export const flatTableCaption = () => flatTable().find('caption');
```

**Review by DipperTheDan — APPROVED** [review]

**Review by mkrds — APPROVED** [review]

**Review by Alec-Walker-Sage — APPROVED** [review]

**Pull request merged**